README.md
# resourceful

A Clojure library which enhances
[Compojure’s](https://github.com/weavejester/compojure) routes DSL by adding the macro
`resource` which a developer can use to declare a *resource*, as opposed to a *route*. This allows the code for a Web API to use the native paradigm of the Web: resources.

`resource` expands into a call to Compojure’s `routes`, so it can be used anywhere `routes` can be
used.

This has the following concrete benefits:

* Allows the path to be specified only once even if a resource supports multiple methods
* Adds an OPTIONS route which returns an Allow header (if OPTIONS is not specified)
* Adds an ANY route to return a 405 response for any unsupported method

...and one abstract but important benefit:

* helps developers think in terms of *resources* rather than *routes* by making the *code* about
  *resources* rather than *routes*

## Background

For the story behind Resourceful see [this blog post](http://aviflax.com/post/my-first-open-source-clojure-library-resourceful/).

## Installation

Add the following dependency to your project.clj file:

```clojure
[resourceful "0.1.1"]
```

## Usage

### Requiring the macro `resource`

Add the following line to the `:require` part of your `ns` call:

```clojure
[resourceful :refer [resource]]
```

### Defining Resources

Let’s say you have a simple resource which supports GET and POST.

With “vanilla” Compojure, you might implement it like so:

```clojure
(defroutes app
  (GET "/authors/:author/books" [author]
    (get-books author))
  (POST "/authors/:author/books" [author title]
    (create-book author)
    (get-books author)))
```

...and your resource wouldn’t properly support OPTIONS, PUT, etc.

Whereas with Resourceful, the same resource would be expressed as:

```clojure
(defroutes app
  (resource "Collection of the books of an author"
            "/authors/:author/books"
            (GET [author]
              (get-books author))
            (POST [author title]
              (create-book author)
              (get-books author))))
```

As you can see, the path to the resource only had to be supplied once, and the macro encourages
developers to supply a name for their resource, in accordance with the REST precept that a resource
is really a *concept*.

In addition to the more concise and expressive code, the resulting set of routes are also more
robust and conformant to HTTP than the “vanilla” example the resource will respond properly to
OPTIONS requests and to requests which specify an unsupported method (with a 405).

## Readability advantage

Another advantage is that if your app has multiple resources — and most do — then using `resource`
makes each resource very distinct within the code — whereas when using vanilla Compojure you end up
with just a linear mess of routes. Sure, you could group the routes using comments and blank lines,
but most people don’t bother. With `resource`, the code becomes self-documenting, and the shape of
the code mirrors the shape of your RESTful application/service/API.

## License

Copyright © 2013 Avi Flax

Distributed under the Eclipse Public License, the same as Clojure.


changes.md
# Change History for Resourceful

## 0.1.1 ~ 22 November 2013

* Removed addition of HEAD route (93bf8de951)
* Updated compojure and ring dependencies (aa88092f2e)

## 0.1.0 ~ 10 October 2013

* Initial release!


test/resourceful/test.clj
(ns resourceful.test
  (:use clojure.test)
  (:require [resourceful :refer [resource]]
            [compojure.core :refer [GET HEAD OPTIONS POST]]
            [clojure.java.io :refer [file input-stream]]
            [clojure.string :refer [split]]))

;; TODO: use ring-mocks for more realistic testing

(deftest test-resource
  (testing "a resource with just a GET route should handle HEAD"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body "bar"}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "3"}))))


  (testing "a resource with a String response body should handle HEAD properly"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body "bar"}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "3"}))))


  (testing "a resource with a InputStream response body should handle HEAD properly"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body (input-stream (.getBytes "bar"))}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "3"}))))


  (testing "a resource with a File response body should handle HEAD properly"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body (file "resources/test/fixed_length_3.txt")}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "3"}))))


  (testing "a resource with a Seq response body should handle HEAD properly"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body (seq ["f" "o" "o"])}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "3"}))))


  (testing "a resource wherein GET sets Content-Length should return that value on HEAD"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain", "Content-Length" "613"}
                             :body "WHATEVER"}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "613"}))))




  (testing "when a GET route does NOT set Content-Length, and the body is an non-standard
            type, the header Content-Length should not be present in the response"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body ["foo"]}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"}))))



  (testing "when a resource supplies its own HEAD handler, it should be used"
    (let [handler (resource "foo" "/"
                    (HEAD [] {:status 200
                              :headers {"Content-Type" "application/yourmom"
                                        "Content-Length" "1337"
                                        "Cheese" "Moldy"}})
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain", "Content-Length" "613"}
                             :body "WHATEVER"}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "application/yourmom"
                             "Content-Length" "1337"
                             "Cheese" "Moldy"}))))


  (testing "when a resource doesn’t supply a GET route, it shouldn’t support HEAD"
    (let [handler (resource "foo" "/"
                    (POST [] {:status 200
                              :headers {"Content-Type" "text/plain", "Content-Length" "8"}
                              :body "Success!"}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 405))
      (is (.contains (get-in res [:headers "Content-Type"]) "text/plain"))
      (is (= (get-in res [:headers "Content-Length"]) "18"))
      (let [res-allow-header (get-in res [:headers "Allow"])
            res-allow-methods (set (split res-allow-header #", "))]
        (is (= res-allow-methods #{"POST" "OPTIONS"})))
      (is (= (:body res "Method Not Allowed")))))


  (testing "when a resource doesn’t supply a POST route, a POST request should return 405"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                              :headers {"Content-Type" "text/plain", "Content-Length" "3"}
                              :body "foo"}))
          req {:uri "/"
               :request-method :post
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 405))
      (is (.contains (get-in res [:headers "Content-Type"]) "text/plain"))
      (is (= (get-in res [:headers "Content-Length"]) "18"))
      (let [res-allow-header (get-in res [:headers "Allow"])
            res-allow-methods (set (split res-allow-header #", "))]
        (is (= res-allow-methods #{"GET" "HEAD" "OPTIONS"})))
      (is (= (:body res "Method Not Allowed")))))


  (testing "a resource should support OPTIONS"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body "bar"})
                    (POST [] {:status 200
                              :headers {"Content-Type" "text/plain", "Content-Length" "8"}
                              :body "Success!"}))
          req {:uri "/"
               :request-method :options}
          res (handler req)]
      (is (= (:status res) 204))
      (is (empty? (:body res)))
      (let [res-allow-header (get-in res [:headers "Allow"])
            res-allow-methods (set (split res-allow-header #", "))]
        (is (= res-allow-methods #{"GET" "HEAD" "POST" "OPTIONS"})))))

  (testing "a response to an OPTIONS request should include the header `Description`"
    (let [description "A Collection of Snark"
          handler (resource description "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body "bar"}))
          req {:uri "/"
               :request-method :options}
          res (handler req)]
      (is (= (:status res) 204))
      (is (empty? (:body res)))
      (let [res-description-header (get-in res [:headers "Description"])]
        (is (= res-description-header description))))))


test/resourceful/core_test.clj
(ns resourceful.core-test
  (:use clojure.test)
  (:require [resourceful :refer [resource]]
            [compojure.core :refer [GET HEAD OPTIONS POST]]
            [clojure.java.io :refer [file input-stream]]
            [clojure.string :refer [split]]))

;; TODO: use ring-mocks for more realistic testing

(deftest test-resource
  (testing "a resource with just a GET route should handle HEAD"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body "bar"}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "3"}))))


  (testing "a resource with a String response body should handle HEAD properly"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body "bar"}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "3"}))))


  (testing "a resource with a InputStream response body should handle HEAD properly"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body (input-stream (.getBytes "bar"))}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "3"}))))


  (testing "a resource with a File response body should handle HEAD properly"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body (file "resources/test/fixed_length_3.txt")}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "3"}))))


  (testing "a resource with a Seq response body should handle HEAD properly"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body (seq ["f" "o" "o"])}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "3"}))))


  (testing "a resource wherein GET sets Content-Length should return that value on HEAD"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain", "Content-Length" "613"}
                             :body "WHATEVER"}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"
                             "Content-Length" "613"}))))




  (testing "when a GET route does NOT set Content-Length, and the body is an non-standard
            type, the header Content-Length should not be present in the response"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body ["foo"]}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "text/plain"}))))



  (testing "when a resource supplies its own HEAD handler, it should be used"
    (let [handler (resource "foo" "/"
                    (HEAD [] {:status 200
                              :headers {"Content-Type" "application/yourmom"
                                        "Content-Length" "1337"
                                        "Cheese" "Moldy"}})
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain", "Content-Length" "613"}
                             :body "WHATEVER"}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 200))
      (is (empty? (:body res)))
      (is (= (:headers res) {"Content-Type" "application/yourmom"
                             "Content-Length" "1337"
                             "Cheese" "Moldy"}))))


  (testing "when a resource doesn’t supply a GET route, it shouldn’t support HEAD"
    (let [handler (resource "foo" "/"
                    (POST [] {:status 200
                              :headers {"Content-Type" "text/plain", "Content-Length" "8"}
                              :body "Success!"}))
          req {:uri "/"
               :request-method :head
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 405))
      (is (.contains (get-in res [:headers "Content-Type"]) "text/plain"))
      (is (= (get-in res [:headers "Content-Length"]) "18"))
      (let [res-allow-header (get-in res [:headers "Allow"])
            res-allow-methods (set (split res-allow-header #", "))]
        (is (= res-allow-methods #{"POST" "OPTIONS"})))
      (is (= (:body res "Method Not Allowed")))))


  (testing "when a resource doesn’t supply a POST route, a POST request should return 405"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                              :headers {"Content-Type" "text/plain", "Content-Length" "3"}
                              :body "foo"}))
          req {:uri "/"
               :request-method :post
               :headers {"accept" "*/*"}}
          res (handler req)]
      (is (= (:status res) 405))
      (is (.contains (get-in res [:headers "Content-Type"]) "text/plain"))
      (is (= (get-in res [:headers "Content-Length"]) "18"))
      (let [res-allow-header (get-in res [:headers "Allow"])
            res-allow-methods (set (split res-allow-header #", "))]
        (is (= res-allow-methods #{"GET" "HEAD" "OPTIONS"})))
      (is (= (:body res "Method Not Allowed")))))


  (testing "a resource should support OPTIONS"
    (let [handler (resource "foo" "/"
                    (GET [] {:status 200
                             :headers {"Content-Type" "text/plain"}
                             :body "bar"})
                    (POST [] {:status 200
                              :headers {"Content-Type" "text/plain", "Content-Length" "8"}
                              :body "Success!"}))
          req {:uri "/"
               :request-method :options}
          res (handler req)]
      (is (= (:status res) 204))
      (is (empty? (:body res)))
      (let [res-allow-header (get-in res [:headers "Allow"])
            res-allow-methods (set (split res-allow-header #", "))]
        (is (= res-allow-methods #{"GET" "HEAD" "POST" "OPTIONS"}))))))


src/resourceful.clj
(ns resourceful
  (:require [clojure.string :refer [join]]
            [compojure.core :refer [ANY HEAD OPTIONS routes]]
            [ring.util.response :refer [header]]))

(defmacro resource
  "Provides more concise and more RESTful alternative to Compojure’s `routes`.

  Specifically:

   * Allows the path to be specified only once even if a resource supports multiple methods
   * Adds an OPTIONS route which returns an Allow header, if OPTIONS is not specified
   * Adds an ANY route to return a 405 response for any unsupported method

   `methods` should be 1–N standard compojure route forms, except with the path omitted.

   Expands into a call to `routes`, so can be used anywhere `routes` can be used.

   For example:

   (resource \"Collection of the books of an author\"
             \"/authors/:author/books\"
             (GET [author] (get-books author))
             (POST [author title] (create-book author) (get-books author)))"
  [description path & methods]
  (let [method-symbols (set (map first methods))
        allowed (->> (map str method-symbols)
                     (concat ["OPTIONS" (when (or (method-symbols 'HEAD)
                                                  (method-symbols 'GET))
                                              "HEAD")] ,,,)
                     (remove nil? ,,,)
                     (join ", " ,,,))]
    `(routes
      ;; Building a list “manually” using concat (as opposed to just unquote-splicing)
      ;; because the “when” forms can produce nil values which must be filtered out of the list
      ~@(remove nil?
          (concat
            ;; output the provided methods/routes
            ;; the method-symbols will be output exactly as provided, so if they were
            ;; provided unqualified, they’ll be output unqualified. I think this is OK
            ;; because the calling NS should have referred the symbols anyway.
            (map (fn [[method-symbol bindings & exprs]]
                   `(~method-symbol ~path ~bindings ~@exprs))
                 methods)

            [
             ;; output OPTIONS, if it isn’t already provided
             (when-not (method-symbols 'OPTIONS)
               `(OPTIONS ~path [] {:status 204
                                   :headers {"Allow" ~allowed
                                             "Description" ~description}
                                   :body nil}))

             ;; output an ANY route to return a 405 for any unsupported method
             (when-not (method-symbols 'ANY)
               `(ANY ~path [] {:status 405
                               :headers {"Allow" ~allowed
                                         "Content-Type" "text/plain;charset=UTF-8"
                                         "Content-Length" "18"}
                               :body "Method Not Allowed"}))
            ])))))
